File: src/ns/track-view.ts
import type { Ableton } from "../index.js";
import { type AnyDevice, type RawDevice, wrapDevice } from "./device.js";
import { Namespace } from "./index.js";

export type DeviceInsertMode = "default" | "selected_left" | "selected_right";

export interface GettableProperties {
  // device_insert_mode: DeviceInsertMode; – for some reason, Live returns a boolean here
  is_collapsed: boolean;
  selected_device: RawDevice | null;
}

export interface TransformedProperties {
  selected_device: AnyDevice | null;
}

export interface SettableProperties {
  device_insert_mode: DeviceInsertMode;
  is_collapsed: boolean;
}

export interface ObservableProperties {
  // device_insert_mode: DeviceInsertMode;
  is_collapsed: boolean;
  selected_device: RawDevice | null;
}

export class TrackView extends Namespace<
  GettableProperties,
  TransformedProperties,
  SettableProperties,
  ObservableProperties
> {
  constructor(ableton: Ableton, nsid: string) {
    super(ableton, "track-view", nsid);

    this.transformers = {
      selected_device: (device) =>
        device ? wrapDevice(ableton, device) : null,
    };

    this.cachedProps = {
      selected_device: true,
    };
  }

  /**
   * Selects the track's instrument if it has one.
   */
  public async selectInstrument() {
    return this.sendCommand("select_instrument");
  }
}

File: midi-script/Namespaces/TrackView.py
from __future__ import absolute_import

import Live

from .Device import Device
from .Interface import Interface


class TrackView(Interface):
    def __init__(self, c_instance, socket):
        super(TrackView, self).__init__(c_instance, socket)

    def get_ns(self, nsid):
        return Interface.get_obj(nsid).view

    def get_selected_device(self, ns):
        return Device.serialize_device(ns.selected_device)

    def set_device_insert_mode(self, ns, name):
        ns.device_insert_mode = getattr(Live.Track.DeviceInsertMode, str(name))
